### PR TITLE
Adjustments to FluidSynth handling and MIDI conf settings

### PR DESCRIPTION
**Pull request by dreamer**
Detailed explanation in each commit separately.

### CODE REVIEW

File: src/midi/midi.cpp [hunk 1]
```diff
@@ -252,18 +254,24 @@ class MIDI : public Module_base {
 		LOG_MSG("MIDI: Can't find device: %s, using default handler.",
 		        dev.c_str());
 getdefault:
-		handler=handler_list;
-		while (handler) {
+		for (handler = handler_list; handler; handler = handler->next) {
```
Comment (kcgen) [inline src/midi/midi.cpp:257]: This loop refactoring is very easy to read, and I like how the loop management is together. Very minor but a nice touch!

File: src/dosbox.cpp [hunk 2]
```diff
@@ -624,20 +619,42 @@ void DOSBOX_Init(void) {
...
-	                  "When using a Roland MT-32 rev. 0 as midi output device, some games may require a delay in order to prevent 'buffer overflow' issues.\n"
-	                  "In that case, add 'delaysysex', for example: midiconfig=2 delaysysex\n"
-	                  "See the README/Manual for more details.");
+	        "Use 'auto' to pick the first working external device.");
+
+	pstring = secprop->Add_string("midiconfig", when_idle, "");
+	pstring->Set_help(
+	        "Special configuration options for the device driver.\n"
+	        "This is usually the id or part of the name of the device you\n"
+	        "want to use (find the id/name with DOS command 'mixer /listmidi').\n"
```
Comment (kcgen) [inline src/dosbox.cpp:634]: I like the last sentence describing how to how discover the value by running the mixer command.

The first sentence adds more confusion, so we should try to reword it. 

- What's so "special" about this configuration value versus other non-special conf values? (we should drop "Special").

- It mentions a "Device driver", yet I believe all of the MIDI players are software (they're not hardware nor are there device-drivers for the player). 

My understanding is that this value tells DOSBox how to interface with selected MIDI player. Is that right?

File: src/dosbox.cpp [hunk 3]
```diff
@@ -624,20 +619,42 @@ void DOSBOX_Init(void) {
...
+#if C_FLUIDSYNTH
+	        "- When using built-in FluidSynth, this option has no effect.\n"
+	        "  Use section 'fluidsynth' below instead.\n"
```
Comment (kcgen) [inline src/dosbox.cpp:637]: When using the built-in FluidSynth (mididevice = fluidsynth), this option has no effect. Instead, configure FluidSynth using the dedicated [fluidsynth] section below.

File: src/midi/midi.cpp [hunk 4]
```diff
@@ -252,18 +254,24 @@ class MIDI : public Module_base {
...
-				LOG_MSG("MIDI: Opened device: %s",
-				        handler->GetName());
+				LOG_MSG("MIDI: Opened device: %s", name.c_str());
 				return;
 			}
-			handler=handler->next;
 		}
-		/* This shouldn't be possible */
+		assert(handler != nullptr);
+		assert(handler->GetName() == "none"s); // -V522
```
Comment (kcgen) [inline src/midi/midi.cpp:274]: The left hand side returns a char*, so we're not able to use std::string == operator from the "none"s string literal.

Swapping them, assert("none"s == handler->GetName()); should work but it less intuitive to read.

Maybe cast/wrap the LHS in a std::string?

Comment (kcgen) [reply src/midi/midi.cpp:274]: If `-V522` is a PVS suppression, we can add it to the `.pvs-suppress` file:
The PVS report contains a `general/suppression-list.json` file that has snippets for each detected issue. You can paste the snippet into `.pvs-suppress` and should take care of it (even if the code is moved around in future changes; it's pretty smart).

File: src/dosbox.cpp (hunk 2 above)
Comment (dreamer) [reply src/dosbox.cpp:634]: There's nothing special about it, I agree.

MIDI players can be hardware (e.g. external keyboard plugged in or modern sound card (however rare they are)).

This value behaves differently depending on what mididevice is selected - usually it selects which available MIDI device to use.

File: src/midi/midi.cpp (hunk 4 above)
Comment (dreamer) [reply src/midi/midi.cpp]: Yeah, V522 was PVS warning - about the `handler` possibly being null (despite the assertion in line earlier). I combined these two assertions into one, and this resolved the warnings as well :)

File: src/dosbox.cpp (hunk 2 above)
Comment (kcgen) [reply src/dosbox.cpp]: >  depending on what mididevice is selected - usually it selects which available MIDI device to use.

There is some circular logic here. (_Depending on the midi device, it selects the midi device_).

Dosbox can send MIDI sequences through a selected **interface** (ALSA, CoreAudio, and WindowsMIDI/win32), which may play the MIDI itself in the case of CoreAudio (when a path to an SF2 is provided), or route the sequence to one or more players (physical mt32, Qsynth, Timidity, munt, ...) that are connected to the interface, via a **connection identifier**.

We want the user to choose their _MIDI interface_ and either _connection identifier_ or _SF2_ in the case of CoreAudio.

Ideally we could rename these settings.. but that would hurt backward compatibility. Maybe we can clarify the descriptions.

File: src/dosbox.cpp [hunk 5]
```diff
@@ -624,20 +619,43 @@ void DOSBOX_Init(void) {
...
-	                  "(find the id/name with mixer/listmidi).\n"
-	                  "Or in the case of coreaudio, you can specify a soundfont here.\n"
-	                  "When using a Roland MT-32 rev. 0 as midi output device, some games may require a delay in order to prevent 'buffer overflow' issues.\n"
-	                  "In that case, add 'delaysysex', for example: midiconfig=2 delaysysex\n"
-	                  "See the README/Manual for more details.");
+	        "Use 'auto' to pick the first working external device.");
+
+	pstring = secprop->Add_string("midiconfig", when_idle, "");
+	pstring->Set_help(
+	        "Configuration options for the selected 'mididevice'.\n"
```
Comment (kcgen) [inline src/dosbox.cpp:632]: The wording is somewhat circular: _Depending on the midi device, it selects the midi device_.

Here's my understanding (and possible wording.. albeit a bit long):

Dosbox can send MIDI sequences through a selected **interface** (ALSA, CoreAudio, and WindowsMIDI/win32), which may play the MIDI itself in the case of CoreAudio (when a path to an SF2 is provided), or route the sequence to one or more players (physical mt32, Qsynth, Timidity, munt, ...) that are connected to the interface, via a **connection identifier**.

We want the user to choose their _MIDI interface_ and either _connection identifier_ or _SF2_ in the case of CoreAudio.

Ideally we could rename these settings.. but that would hurt backward compatibility. Maybe we can clarify the descriptions.

Comment (kcgen) [reply src/dosbox.cpp:632]: Thanks for the updates, @dreamer!

Will take another try tomorrow at this midiconfig text with fresh eyes. Don't want us burning too much time on it.

**Comment by kcgen** [comment]
Thanks for the updates, @dreamer!

Will take another try tomorrow at this midiconfig text with fresh eyes, so we can close this out.

File: src/dosbox.cpp (hunk 5 above)
Comment (dreamer) [reply src/dosbox.cpp:632]: Yes, it *is* too long. Every sentence added means someone will not read it at all. Also "connection identifier" would be inventing yet another, new term: on Linux it's: MIDI sequencer port (but I have no idea about nomenclature on other OSes).

Let's keep this as brief as possible

How about:

    midiconfig: Configuration options for the selected MIDI interface.
                This is usually the id or name of the MIDI synthesiser you want
                to use (find the id/name with DOS command 'mixer /listmidi').

Comment (kcgen) [reply src/dosbox.cpp:632]: You nailed it with "Configuration options for the selected MIDI interface."! Thanks @dreamer.

**Review by kcgen — APPROVED** [review]
(with your suggested first sentence for the `midiconfig` description). 

This is a very welcome improvement @dreamer!

**Comment by dreamer** [comment]
@kcgen Thanks for help :)

One more task from https://github.com/dosbox-staging/dosbox-staging/projects/4 done, only 14 more to go…

**Comment by dreamer** [comment]
@grapeli I definitely want to have automatic discovery - I implemented it once already for Boxtron and intent to port it to DOSBox Staging (it's a bit more complicated than hardcoding port 128:0).

We have no guarantee that 128:0 is open for writing, but testing port 65:0 and 128:0 is better than (inherited from upstream) 65:0 and 17:0.

**Comment by grapeli** [comment]
@dreamer 
That's what I mean. After all, this is a cosmetic fix (and the effectiveness is huge). **Don't use the default port?**
Auto-discovering is probably troublesome to implement (a waste of time).

**Comment by dreamer** [comment]
@grapeli No, auto-discovery is not that hard to implement :) ALSA provides functions to iterate through the open ports, we just need to pick the first one available for writing.

I just tried replacing 17:0 with 128:0 and it caused a new problem… when an external synth is not started before DOSBox (which will be the usual case), the 128:0 is DOSBOX port and current implementation simply plugs dosbox output into dosbox input in this case…

This code is generally *awful* and I don't want to touch it before 0.76.0 release as we are trying to stabilize now - added a new task for 0.77.0: https://github.com/dosbox-staging/dosbox-staging/projects/9

**Comment by grapeli** [comment]
> I just tried replacing 17:0 with 128:0 and it caused a new problem… when an external synth is not started before DOSBox (which will be the usual case), the 128:0 is DOSBOX port and current implementation simply plugs dosbox output into dosbox input in this case…

Exactly as you described, I checked it.

**Comment by grapeli** [comment]
Overall, the same behavior as the current code.
Set to hard client:port `midiconfig=128:0`. This setting is used by most of the users running software synthesizers like fluidsynth, timidity, adlmidi (munt?). Does anyone use other solutions these days?

Anyway, in a normally behaving program, a connection could be established (from outside).
`aconnect 128:0 129:0`

I saw how it works in ScummVM. Correct (automatic).
https://github.com/scummvm/scummvm/blob/master/backends/midi/alsa.cpp
I suspect that RetroArch works similarly (although I have not checked).
https://github.com/libretro/RetroArch/blob/master/midi/drivers/alsa_midi.c

**Comment by dreamer** [comment]
I do have correct, automatic connection in Boxtron - will just port the behaviour from there to dosbox-staging :)

**Comment by grapeli** [comment]
> I do have correct, automatic connection in Boxtron - will just port the behaviour from there to dosbox-staging :)

Will not work on the kernel with `CONFIG_SND_PROC_FS is not set`.
Anyone can create a single-line shell wrapper to overcome dosbox's shortcomings.

**Comment by dreamer** [comment]
1. Boxtron is not a shell script
2. In Boxtron I parsed procfs because it's extremely hard to create python C bindings to alsalib (due to excessive macro usage in ALSA) - I needed to make decision: good, simple solution that will work for 99% of users or better, complicated solution that will work for 100% users. Guess what I picked :)
3. Even with procfs implementation, we generally had no issues. I know no Linux distro that does not offer sequencer list over procfs - but there were problem for Flatpak/Snap users
4. When porting the solution to dosbox-staging I won't need to worry about C bindings in Python, I am able to use ALSA interface directly…

So no worries, we will fix it - I just don't want to work on it while we are stabilizing and testing 0.76.0 release. If you want to see it in faster - then I'll be happy to see PR.

**Comment by grapeli** [comment]
> 1. Boxtron is not a shell script

Who thinks so? Definitely not me. I distinguish Bash from Python.

> 3. Even with procfs implementation, we generally had no issues. I know no Linux distro that does not offer sequencer list over procfs - but there were problem for Flatpak/Snap users

I knew such. They added 10 days ago. I mean, I don't know now. Very nice distribution. I recommend.
https://github.com/sabotage-linux/sabotage/commit/9608e1f2de4e4785a8d3edabe67df75934051c39

Flatpak? Sorry. I will not run binary code from outside the repository (of unknown origin) that will be able to access my home directory.

**Comment by dreamer** [comment]
@grapeli I think you need to chill down today - leave comments only if they are adding something useful to the discussion.

Boxtron is packaged in Flatpak as Steam extension (if Flatpak is your thing), but I see no way it is adding anything relevant to this, already merged-in, PR.

**Pull request merged**